# Notebook 02 — Data Cleaning

## Objectives
- Fix `TotalCharges` — convert to float and impute blank values with the median
- Drop `customerID` (non-predictive identifier)
- Encode the Churn target: Yes = 1, No = 0
- Save the cleaned dataset

## Inputs
- `outputs/datasets/collection/telco_churn_raw.csv`

## Outputs
- `outputs/datasets/collection/telco_churn_cleaned.csv`

---

## 1. Change Working Directory

In [1]:
import os

os.chdir('..') if os.path.basename(os.getcwd()) == 'jupyter_notebooks' else None
print('Working directory:', os.getcwd())

Working directory: /workspaces/customer-churn-predictor


## 2. Import Libraries

In [2]:
import pandas as pd
import numpy as np

## 3. Load Raw Dataset

In [3]:
df = pd.read_csv('outputs/datasets/collection/telco_churn_raw.csv')
print(f'Shape: {df.shape}')
df.head(3)

Shape: (7043, 21)


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes


## 4. Fix TotalCharges

Blank strings belong to customers with zero tenure who have not yet received a bill.
We replace them with NaN, cast to float, then impute with the column median.

In [4]:
# Replace blank strings, cast to numeric
df['TotalCharges'] = df['TotalCharges'].replace(' ', np.nan)
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

# Impute missing values with the median
median_val = df['TotalCharges'].median()
df['TotalCharges'] = df['TotalCharges'].fillna(median_val)

print(f'Imputed {df["TotalCharges"].isnull().sum()} nulls with median {median_val:.2f}')

Imputed 0 nulls with median 1397.47


## 5. Drop customerID

In [5]:
# customerID is a unique identifier with no predictive value
df = df.drop(columns=['customerID'])
print('Shape after drop:', df.shape)

Shape after drop: (7043, 20)


## 6. Encode Target Variable

In [6]:
df['Churn'] = df['Churn'].map({'Yes': 1, 'No': 0})
print(df['Churn'].value_counts())

Churn
0    5174
1    1869
Name: count, dtype: int64


## 7. Validate Cleaned Dataset

In [7]:
print('Shape:', df.shape)
print('\nDtypes:')
print(df.dtypes)
print('\nMissing values:')
print(df.isnull().sum())

Shape: (7043, 20)

Dtypes:
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                 int64
dtype: object

Missing values:
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyC

## 8. Save Cleaned Dataset

In [8]:
df.to_csv('outputs/datasets/collection/telco_churn_cleaned.csv', index=False)
print('Saved: telco_churn_cleaned.csv')

Saved: telco_churn_cleaned.csv


---
## Conclusions
- 11 blank strings in `TotalCharges` were imputed with the median.
- `customerID` was dropped as it carries no predictive signal.
- `Churn` was label-encoded: No = 0, Yes = 1.
- Cleaned dataset has 7,043 rows × 20 columns with no missing values.